sycl/source/detail/graph_impl.cpp

```diff
@@ -346,6 +356,7 @@ void exec_graph_impl::createURCommandBuffers(sycl::device Device) {
 }
 
 exec_graph_impl::~exec_graph_impl() {
+  MMutex.lock();
```

Comment:
Can we use RAII for some of these `lock()`/`unlock()` usage combinations so that the mutex gets automatically unlocked when a variable goes out of scope, something like [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard)

Reply:
std::lock_guard does not enable to do lock_shared and unlock_shared. I therefore prefer not to mix std::lock_guard with lock()/unlock() coding style in the same file.
A solution might be to code a custom  lock_guard class that manages shared_lock?

Reply:
You can use `std::shared_lock` and `std::unique_lock` for this purpose. Quite common to do something like this in the class declaration:

```c++
using ReadLock = std::shared_lock<std::shared_mutex>;
using WriteLock = std::unique_lock<std::shared_mutex>;
```
Then you can use them to get an RAII style lock:

```c++
void foo(){
  ReadLock lock(MMutex);
  ...
}
```

Reply:
I updated the implementation this way.